cumulusci/core/github.py

```diff
@@ -268,3 +303,131 @@ def get_version_id_from_tag(repo: Repository, tag_name: str) -> str:
+    else:
+        scope_error_msg = check_github_scopes(exc)
+        sso_error_msg = check_github_sso_auth(exc)
+        user_warning = scope_error_msg + sso_error_msg
+
+    return user_warning
+
+
+def check_github_scopes(exc: ResponseError) -> str:
+    """
+    Parse github3 ResponseError headers for the correct scopes and return a
+    warning if the user is missing.
+
+    @param exc: The exception to process
+    @returns: The formatted exception string
+    """
+
+    user_warning = ""
+
+    has_wrong_status_code = exc.response.status_code not in (403, 404)
+    if has_wrong_status_code:
+        return user_warning
+
+    token_scopes = get_oauth_scopes(exc.response)
+
+    # Gist resource won't return X-Accepted-OAuth-Scopes for some reason, so this
+    # string might be `None`; we discard the empty string if so.
+    accepted_scopes = exc.response.headers.get("X-Accepted-OAuth-Scopes") or ""
+    accepted_scopes = set(accepted_scopes.split(", "))
+    accepted_scopes.discard("")
+
+    request_url = urlparse(exc.response.url)
+    if not accepted_scopes and request_url.path == "/gists":
+        accepted_scopes = {"gist"}
+
+    missing_scopes = accepted_scopes.difference(token_scopes)
+    if missing_scopes:
+        user_warning = f"Your token may be missing the following scopes: {', '.join(missing_scopes)}\n"
+        # This assumes we're not on enterprise and 'api.github.com' == request_url.hostname
+        user_warning += "Visit https://github.com/settings/tokens to add them."
+
+    return user_warning
+
+
+def check_github_sso_auth(exc: ResponseError) -> str:
+    """
+    Check ResponseError header for SSO authorization and return a warning if
+    required
+
+    @param exc: The exception to process
+    @returns: The formatted exception string
+    """
+    user_warning = ""
+    headers = exc.response.headers
+
+    if exc.response.status_code != 403 or "X-Github-Sso" not in headers:
+        return user_warning
+
+    sso_header = str(headers["X-Github-Sso"] or "")
+    if sso_header.startswith("required; url="):
+        # In this case the message from github is good enough, but we can help
+        # the user by opening a browser to authorize the token.
+        auth_url = sso_header.split("url=", maxsplit=1)[1]
+        user_warning = f"{exc.message}\n{auth_url}"
+        webbrowser.open(auth_url)
+    elif sso_header.startswith("partial-results"):
+        # In cases where we don't have complete results we get the
+        # partal-results header, so return the organization IDs. This may or
+        # may not be useful without help from us to lookup the org IDs.
+        unauthorized_org_ids = get_sso_disabled_orgs(exc.response)
+        user_warning = f"Results may be incomplete. You have not granted your Personal Access token access to the following organizations: {unauthorized_org_ids}"
+
+    return user_warning
+
+
+def get_sso_disabled_orgs(response: Response) -> list:
```

Comment:
Isn't this type hint syntax limited to Python 3.9+?

Reply:
What's the specific syntax that is in question here, the function argument type and return value? We have this in many other places in the code base, so I'm wondering what is different about this?

Reply:
Lowercase `list` - from [PEP 585](https://www.python.org/dev/peps/pep-0585/). It's supported with a future annotation, though.

Reply:
Interesting, as @davisagli indicated to me that I could use the lowercase `dict` type [here](https://github.com/SFDO-Tooling/CumulusCI/pull/2676#discussion_r662549413).

Reply:
PEP 585 is about using built-ins as a generic (e.g. `list[str]`). Using a built-in as a hint (`-> list:`) is supported in 3.5+. See: [PEP 484](https://www.python.org/dev/peps/pep-0484/#acceptable-type-hints)